File: clients/javascript/__tests__/features/arguments.spec.js
import {default as Invalid} from "../../src/errors/invalid";

require("../support/acceptance");

const Proc = require("../../index");
const client = Proc.connect(process.env.SECRET);

describe("using an argument reference as input in a call", () => {
  test("uses the given argument as input", async () => {
    const proc = client.core.echo(client.argument("foo"));

    expect(await proc.call(undefined, {foo: "foo"})).toEqual("foo");
  });
});

describe("using an argument reference to require an argument in a call", () => {
  let proc = client.type.string.truncate(undefined, {length: client.argument("truncate_to")});

  test("uses the given argument in the call", async () => {
    expect(await proc.call("foo", {truncate_to: 2})).toEqual("fo")
  });

  test("uses the given argument using with", async () => {
    expect(await proc.with(undefined, {truncate_to: 2}).call("foo")).toEqual("fo")
  });

  test("uses complex argument values", async () => {
    expect(await proc.call("foo", {truncate_to: client.core.echo(2)})).toEqual("fo")
  });

  test("fails when the argument is not given", async () => {
    await expect(() => proc.call("foo")).rejects.toThrowError(Invalid);
    await expect(() => proc.call("foo")).rejects.toThrowError("invalid argument `truncate_to' for `type.string.truncate' (is required)");
  });

  test("prefers the given argument", async () => {
    proc = client.type.string.truncate(undefined, {length: client.argument("length")})

    expect(await proc.call("foo", {length: 2})).toEqual("fo");
  });
});

describe("using an argument reference as input in a composition", () => {
  let proc = client.type.string.reverse(client.argument("foo")).compose(client.type.string.truncate(undefined, {length: 2}));

  test("uses the given argument value as input", async () => {
    expect(await proc.call(undefined, {foo: "bar"})).toEqual("ra")
  });

  test("uses the given argument value as input using with", async () => {
    expect(await proc.with(undefined, {foo: "bar"}).call()).toEqual("ra")
  });

  test("uses complex argument values", async () => {
    expect(await proc.call(undefined, {foo: client.core.echo("bar")})).toEqual("ra")
  });
});

describe("using an argument reference to require an argument in a composition", () => {
  let proc = client.type.string.reverse.compose(
    client.type.string.truncate(undefined, {length: client.argument("truncate_to")})
  );

  test("uses the given argument in the composition", async () => {
    expect(await proc.call("foo", {truncate_to: 1})).toEqual("o");
  });

  test("uses the given argument in the composition using with", async () => {
    expect(await proc.with(undefined, {truncate_to: 1}).call("foo")).toEqual("o");
  });

  test("uses complex argument values", async () => {
    expect(await proc.call("foo", {truncate_to: client.core.echo(1)})).toEqual("o");
  });

  test("fails when the argument is not passed", async () => {
    await expect(() => proc.call("foo")).rejects.toThrowError(Invalid);
    await expect(() => proc.call("foo")).rejects.toThrowError("invalid argument `truncate_to' for `type.string.truncate' (is required)");
  });

  test("prefers the given argument", async () => {
    proc = client.type.string.reverse.compose(
      client.type.string.truncate(undefined, {length: client.argument("length")})
    );

    expect(await proc.call("foo", {length: 1})).toEqual("o");
  });
});

describe("passing a complex value to an argument reference in a composition", () => {
  let proc = client.type.string.reverse.compose(
    client.type.string.truncate(undefined, {length: client.argument("truncate_to")})
  );

  test("resolves the complex value correctly", async () => {
    expect(await proc.call("foo", {truncate_to: client.core.echo(1)})).toEqual("o");
  });
});

describe("using an argument reference with a default value in a composition", () => {
  let proc = client.type.string.reverse.compose(
    client.type.string.truncate(undefined, {length: client.argument("truncate_to", {default: 2})})
  );

  test("uses argument references for a composition", async () => {
    expect(await proc.call("foo", {truncate_to: 1})).toEqual("o");
  });

  test("falls back to the default value when not passed", async () => {
    expect(await proc.call("foo")).toEqual("oo");
  });
});

describe("using an argument reference with a complex default value in a composition", () => {
  let echo = client.core.echo(2);
  let proc = client.type.string.reverse.compose(
    client.type.string.truncate(undefined, {length: client.argument("truncate_to", {default: echo})})
  );

  test("falls back to the default value when not passed", async () => {
    expect(await proc.call("foo")).toEqual("oo");
  });
});

describe("coercing an argument reference", () => {
  test("coerces the value correctly", async () => {
    expect(await client.core.echo(client.arg("foo", {type: "integer"})).call(undefined, {foo: "1"})).toEqual(1);
  });
});

describe("using the convenience method to define an argument reference", () => {
  let proc = client.type.string.reverse.compose(
    client.type.string.truncate(undefined, {length: client.arg("truncate_to")})
  );

  test("uses the argument reference", async () => {
    expect(await proc.call("foo", {truncate_to: 1})).toEqual("o");
  });
});

File: clients/ruby/spec/features/arguments_spec.rb
# frozen_string_literal: true

require_relative "support/context/acceptance"

RSpec.describe "ruby client api: arguments" do
  include_context "acceptance"

  describe "using an argument reference as input in a call" do
    let(:proc) {
      client.core.echo(client.argument(:foo))
    }

    it "uses the given argument as input" do
      expect(proc.call(foo: "foo")).to eq("foo")
    end
  end

  describe "using an argument reference to require an argument in a call" do
    let(:proc) {
      client.type.string.truncate(length: client.argument(:truncate_to))
    }

    it "uses the given argument in the call" do
      expect(proc.call("foo", truncate_to: 2)).to eq("fo")
    end

    it "uses the given argument using with" do
      expect(proc.with(truncate_to: 2).call("foo")).to eq("fo")
    end

    it "uses complex argument values" do
      expect(proc.call("foo", truncate_to: client.core.echo(2))).to eq("fo")
    end

    it "fails when the argument is not given" do
      expect {
        proc.call("foo")
      }.to raise_error(Proc::Invalid, "invalid argument `truncate_to' for `type.string.truncate' (is required)")
    end

    context "argument name conflicts with the argument needed by the call" do
      let(:proc) {
        client.type.string.truncate(length: client.argument(:length))
      }

      it "uses the given argument in the call" do
        expect(proc.call("foo", length: 2)).to eq("fo")
      end
    end
  end

  describe "using an argument reference as input in a composition" do
    let(:composition) {
      client.type.string.reverse(client.argument(:foo)) >> client.type.string.truncate(length: 2)
    }

    it "uses the given argument value as input" do
      expect(composition.call(foo: "bar")).to eq("ra")
    end

    it "uses the given argument value as input using with" do
      expect(composition.with(foo: "bar").call).to eq("ra")
    end

    it "uses complex argument values" do
      expect(composition.call(foo: client.core.echo("bar"))).to eq("ra")
    end
  end

  describe "using an argument reference to require an argument in a composition" do
    let(:composition) {
      client.type.string.reverse >> client.type.string.truncate(length: client.argument(:truncate_to))
    }

    it "uses the given argument in the composition" do
      expect(composition.call("foo", truncate_to: 1)).to eq("o")
    end

    it "uses the given argument in the composition using with" do
      expect(composition.with(truncate_to: 1).call("foo")).to eq("o")
    end

    it "uses complex argument values" do
      expect(composition.call("foo", truncate_to: client.core.echo(1))).to eq("o")
    end

    it "fails when the argument is not passed" do
      expect {
        composition.call("foo")
      }.to raise_error(Proc::Invalid, "invalid argument `truncate_to' for `type.string.truncate' (is required)")
    end

    context "argument name conflicts with the argument needed by the call" do
      let(:composition) {
        client.type.string.reverse >> client.type.string.truncate(length: client.argument(:length))
      }

      it "uses the given argument in the call" do
        expect(composition.call("foo", length: 1)).to eq("o")
      end
    end
  end

  describe "passing a complex value to an argument reference in a composition" do
    let(:composition) {
      client.type.string.reverse >> client.type.string.truncate(length: client.argument(:truncate_to))
    }

    it "resolves the complex value correctly" do
      expect(composition.call("foo", truncate_to: client.core.echo(1))).to eq("o")
    end
  end

  describe "using an argument reference with a default value in a composition" do
    let(:composition) {
      client.type.string.reverse >> client.type.string.truncate(length: client.argument(:truncate_to, default: 2))
    }

    it "uses argument references for a composition" do
      expect(composition.call("foo", truncate_to: 1)).to eq("o")
    end

    it "falls back to the default value when not passed" do
      expect(composition.call("foo")).to eq("oo")
    end
  end

  describe "using an argument reference with a complex default value in a composition" do
    let(:composition) {
      echo = client.core.echo(2)

      client.type.string.reverse >> client.type.string.truncate(length: client.argument(:truncate_to, default: echo))
    }

    it "falls back to the default value when not passed" do
      expect(composition.call("foo")).to eq("oo")
    end
  end

  describe "coercing an argument reference" do
    it "coerces the value correctly" do
      expect(client.core.echo(client.arg(:foo, type: :integer)).call(foo: "1")).to eq(1)
    end
  end

  describe "using the convenience method to define an argument reference" do
    let(:composition) {
      client.type.string.reverse >> client.type.string.truncate(length: client.arg(:truncate_to))
    }

    it "uses the argument references" do
      expect(composition.call("foo", truncate_to: 1)).to eq("o")
    end
  end

  describe "using symbols as arguments for input" do
    let(:proc) {
      client.core.echo(:foo)
    }

    it "uses the given argument as input" do
      expect(proc.call(foo: "foo")).to eq("foo")
    end
  end

  describe "using an argument reference to require an argument in a call" do
    let(:proc) {
      client.type.string.truncate(length: :truncate_to)
    }

    it "uses the given argument in the call" do
      expect(proc.call("foo", truncate_to: 2)).to eq("fo")
    end
  end
end
